### ISSUE TITLE
merge_bn_in() in QuantConv2d

### ISSUE THREAD

**Issue by andylai0212** [Xilinx/brevitas#59]
Hi,
I found that there is merge_bn_in() in QuantConv2d. How can I use it for merging bn factors into QuantConv2d?
Do I just use nn.BatchNorm2d after qnn.QuantConv2d durning training and then merge them in testing phase?  Also, does the scale factor in QuantConv2d changed after merging bn?

**Comment by N2606** [comment]
Hi, @volcacius 
I would like to quantize BN as well. I found that there is a layer BatchNorm2dToQuantScaleBias, it seems that it use mul_add_from_bn() to merge bn factor into weight and bias. And then perform quantization and learn the scale. My question is that:
1. How can I get bn factors at first. Do I need to train the network with nn.BatchNorm2d first, and then replace it with BatchNorm2dToQuantScaleBias and pass the nn.BatchNorm2dfactors to BatchNorm2dToQuantScaleBias as initial?
2. If I directly replace nn.BatchNorm2d with BatchNorm2dToQuantScaleBias. Does BatchNorm2dToQuantScaleBias contains the information of running_mean and  running_var or it only learns the scale and bias without the information of running_mean and  running_var.

**Comment by volcacius** [comment]
Hello,

Dealing with batch norm is closely related with scale factors and how you plan to implement them in your inference pipeline. There are a lot of alternatives.

If you are using floating point scale factors for your quantization (weight_restrict_scaling_type == 'FP' or 'LOG_FP'), you could keep your standard floating point nn.BatchNorm2d during quantized retraining, and then merge it with the floating-point scale factors. You would still have the bias unquantized, but usually direct quantization of bias is good enough, or you could implement them (both scale and bias) with thresholds like FINN does, and get bit accurate results at inference time.

Regarding BatchNorm2dToQuantScaleBias, what it does is to replace BatchNorm2d with only two learned parameters, a scale and a bias, that are initialized by merging running_mean, running_var, weight and bias of nn.BatchNorm2d. So to answer question (1) and (2), this is designed to replace a pretrained QuantConv2d, no runtime statistics are collected anymore, it's just y = x * scale + bias. What it does it allow you to model fixed point inference of batch norm at training time. It makes sense if for example you are combining it with power of two scale factors (weight_restrict_scaling_type == 'POWER_OF_TWO') for modelling complete fixed point inference.

Another scenario would be to merge batch norm into the weights of QuantConv2d before quantized retraining. This is necessary if for example you are retraining with tensor-wise scale factors, or if you want to completely get rid of batch norm when training with power of two scale factors.
To answer the original question, merging batch into QuantConv2d after training is different from merging it before training. Quantization is a non linear operation, your scale factors would change. For 8 bit in many cases you wouldn't see much difference: for example, our recently release 8 bit QuartzNet with per-tensor scaling is the result of directly merging batch norm into QuantConv2d (trained with per-tensor scaling), without any extra retraining. At lower precision it makes a difference, and you would need extra retraining. 

Alessandro

**Comment by N2606** [comment]
Hi, @volcacius 
Thanks for the elaborate replies. This is very helpful. I am a little bit confused about:

>  So to answer question (1) and (2), this is designed to replace a pretrained QuantConv2d, no runtime statistics are collected anymore, it's just y = x * scale + bias. What it does it allow you to model fixed point inference of batch norm at training time. 

Did it mean to replace a "pretrained BatchNorm2d"?

> It makes sense if for example you are combining it with power of two scale factors (weight_restrict_scaling_type == 'POWER_OF_TWO') for modelling complete fixed point inference.

Did it mean if I have both QuantConv2d and BatchNorm2dToQuantScaleBias with power of two scale factors, then I can model complete fixed point inference?

**Comment by volcacius** [comment]
1) Yes my bad, i meant BatchNorm2d.
2) That's one way to do it. I wouldn't say it's the best one accuracy wise, but you get stricter modelling of the inference data path at training time for a target fixed point hardware.

Alessandro
